packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/DatasetController.java

```diff
@@ -147,19 +147,39 @@ public ResponseEntity<Dataset> getDataset(
 		);
 
 		try {
-			final Optional<Dataset> dataset = datasetService.getAsset(id, permission);
+			Dataset dataset = datasetService
+				.getAsset(id, permission)
+				.orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, messages.get("dataset.not-found")));
 
-			if (dataset.isEmpty()) {
-				return ResponseEntity.noContent().build();
-			}
 			// GETs not associated to a projectId cannot read private or temporary assets
+			if (permission.equals(Schema.Permission.NONE) && (!dataset.getPublicAsset() || dataset.getTemporary())) {
+				throw new ResponseStatusException(HttpStatus.FORBIDDEN, messages.get("rebac.unauthorized-read"));
+			}
+
+			// If the user as write permission, and the stats are not present, calculate them
 			if (
-				permission.equals(Schema.Permission.NONE) && (!dataset.get().getPublicAsset() || dataset.get().getTemporary())
+				permission.equals(Schema.Permission.WRITE) &&
+				dataset.getColumns().stream().anyMatch(column -> column.getStats() == null)
 			) {
-				throw new ResponseStatusException(HttpStatus.FORBIDDEN, messages.get("rebac.unauthorized-read"));
+				// Calculate the statistics for the columns
+				try {
+					final PresignedURL datasetUrl = datasetService
+						.getDownloadUrl(dataset.getId(), dataset.getFileNames().get(0))
+						.orElseThrow(() -> new IllegalArgumentException(messages.get("dataset.download.url.not.found")));
+
+					datasetStatistics.add(dataset, datasetUrl);
+				} catch (final Exception e) {
+					log.error("Error calculating statistics for dataset {}", dataset.getId(), e);
+				}
+
+				// Update and fetch updated dataset
+				datasetService.updateAsset(dataset, projectId, Schema.Permission.WRITE);
+				dataset = datasetService
+					.getAsset(id, permission)
+					.orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, messages.get("dataset.not-found")));
```

Comment:
Because this is just for stats generation (which is sketchy at best here.....), do we want to throw an exception here and fail the request?

Reply:
Good point, updated to be non-breaking